api/firstbook2.py

```diff
@@ -0,0 +1,196 @@
+        # which is also the authorization identifier.
+        return PatronData(
+            permanent_id=username,
+            authorization_identifier=username,
+        )
+
+    # End implementation of BasicAuthenticationProvider abstract methods.
+
+    def remote_pin_test(self, barcode, pin):
+        jwt = self.jwt(barcode, pin)
+        url = self.root + jwt
+        try:
+            response = self.request(url)
+        except requests.exceptions.ConnectionError, e:
+            raise RemoteInitiatedServerError(
+                str(e.message),
+                self.NAME
+            )
+        if response.status_code != 200:
+            msg = "Got unexpected response code %d. Content: %s" % (
+                response.status_code, response.content
+            )
+            raise RemoteInitiatedServerError(msg, self.NAME)
+        if self.SUCCESS_MESSAGE in response.content:
+            return True
+        return False
+
+    def jwt(self, barcode, pin):
+        """Create and sign a JWT with the payload expected by the
+        First Book API.
+        """
+        now = str(int(time.time()))
+        payload = dict(
+            barcode=barcode,
+            pin=pin,
+            iat=now,
+        )
+        return jwt.encode(payload, self.secret, algorithm=self.ALGORITHM)
+
+    def request(self, url):
+        """Make an HTTP request.
+
+        Defined solely so it can be overridden in the mock.
+        """
+        return requests.get(url)
+
+
+class MockFirstBookResponse(object):
+
+    def __init__(self, status_code, content):
+        self.status_code = status_code
+        self.content = content
+
+class MockFirstBookAuthenticationAPI(FirstBookAuthenticationAPI):
+
+    SUCCESS = '"Valid Code Pin Pair"'
+    FAILURE = '{"code":404,"message":"Access Code Pin Pair not found"}'
+
+    def __init__(self, library, integration, valid={}, bad_connection=False,
+                 failure_status_code=None):
+        super(MockFirstBookAuthenticationAPI, self).__init__(
+            library, integration, root="http://example.com/",
+            secret="secret"
+        )
+        self.identifier_re = None
+        self.password_re = None
+
+        self.valid = valid
+        self.bad_connection = bad_connection
+        self.failure_status_code = failure_status_code
+
+        self.request_urls = []
+
+    def request(self, url):
+        self.request_urls.append(url)
+        if self.bad_connection:
+            # Simulate a bad connection.
+            raise requests.exceptions.ConnectionError("Could not connect!")
+        elif self.failure_status_code:
+            # Simulate a server returning an unexpected error code.
+            return MockFirstBookResponse(
+                self.failure_status_code, "Error %s" % self.failure_status_code
+            )
+        parsed = urlparse.urlparse(url)
+        token = parsed.path.split("/")[-1]
+        barcode, pin = self._decode(token)
```

Comment:
The `_decode` call is new, as is `_decode` itself. Previously there was nothing to decode, so we just looked for the barcode and pin in `self.valid`.